network/tools/network-policy-enforcement-latency/pkg/client.go

```diff
@@ -0,0 +1,545 @@
+	workerCount int
+}
+
+type boolMapWithLock struct {
+	mp   map[string]bool
+	lock sync.RWMutex
+}
+
+type timeMapWithLock struct {
+	mp   map[string]time.Time
+	lock sync.RWMutex
+}
+
+type timeWithLock struct {
+	time *time.Time
+	lock sync.RWMutex
+}
+
+type targetSpec struct {
+	ip, name, namespace string
+	startTime           time.Time
+}
+
+type podEvent struct {
+	podName    string
+	isAddEvent bool
+}
+
+// CreateConfig creates a configuration object for test client based on the
+// command flags.
+func CreateConfig() *config {
+	config := config{}
+
+	flag.StringVar(&config.testClientNamespace, "testClientNamespace", "", "The namespace of test client pods")
+	flag.StringVar(&config.targetLabelSelector, "targetLabelSelector", "", "The label selector of target pods to send requests to")
+	flag.StringVar(&config.targetNamespace, "targetNamespace", "", "The namespace of target pods to send requests to")
+	flag.IntVar(&config.targetPort, "targetPort", 0, "The port number of target pods to send requests to")
+	flag.IntVar(&config.expectedTargets, "expectedTargets", 100, "The expected number of target pods to send requests to")
+	flag.StringVar(&config.allowPolicyName, "allowPolicyName", "", "The name of the egress policy that allows traffic from test client pods to target pods")
+	flag.IntVar(&config.metricsPort, "metricsPort", 9154, "The port number where Prometheus metrics are exposed")
+	flag.BoolVar(&config.testPodCreation, "testPodCreation", false, "Specifies whether to test latency for pod creation mode")
+	flag.IntVar(&config.workerCount, "workerCount", utils.DefaultWorkerCount, "The number of workers to process the pod watch events in the workerQueue")
+	flag.Parse()
+
+	return &config
+}
+
+// NewTestClient creates a new test client for the provided config.
+func NewTestClient(config *config, mainStopChan chan os.Signal) *TestClient {
+	return &TestClient{
+		config:       config,
+		MainStopChan: mainStopChan,
+	}
+}
+
+// Run verifies the test client config, starts the metrics server, and runs the
+// network policy enforcement latency test based on the config.
+func (c *TestClient) Run() {
+	if err := c.initialize(); err != nil {
+		klog.Errorf("Failed to initialize: %v", err)
+		c.enterIdleState()
+		return
+	}
+
+	c.metricsServer = metrics.StartMetricsServer(fmt.Sprintf(":%d", c.config.metricsPort))
+	metrics.RegisterMetrics(c.config.testPodCreation)
+	defer func(metricsServer *http.Server, ctx context.Context) {
+		if err := metricsServer.Shutdown(ctx); err != nil {
+			klog.Errorf("Metrics server shutdown failed, error: %v", err)
+		}
+	}(c.metricsServer, context.TODO())
+
+	if c.config.testPodCreation {
+		if err := c.measurePodCreation(); err != nil {
+			klog.Errorf("Pod creation test failed, error: %v", err)
+		}
+	} else {
+		if err := c.measureNetPolicyCreation(); err != nil {
+			klog.Errorf("Pod creation test failed, error: %v", err)
+		}
+	}
```

Comment:
testPodCreation is vague.
Maybe it would be fine if this is additional test you can run, e.g.

if testPodCreation {
   c.measurePodCreation()
}

but in the current form it is not really a boolean whether to test pod creation but rather a flag for picking a mode what to test.

Why can't we just test both of them?
What is the usecase for this flag?

Reply:
I changed the flag name to `podCreationMode`.

I think it’s better to have one test client to run just one of the tests (modes), because of the use case and to keep it simpler.
You can run both at the same time by deploying them in separate pods for each mode.

The flag determines which mode to use. They work differently.
Default is network policy enforcement latency for network policy creation. It lists target pods and tries to reach them (no pod watcher).
When the flag is `True`, it tests network policy enforcement latency for pod creation (creates pod watcher).

Reply:
Split into two applications.